wdl-doc/src/callable.rs

```diff
@@ -0,0 +1,295 @@
+//! HTML generation for WDL callables (workflows and tasks).
+
+pub mod task;
+pub mod workflow;
+
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+
+use maud::Markup;
+use maud::html;
+use wdl_ast::AstToken;
+use wdl_ast::v1::InputSection;
+use wdl_ast::v1::MetadataSection;
+use wdl_ast::v1::MetadataValue;
+use wdl_ast::v1::OutputSection;
+use wdl_ast::v1::ParameterMetadataSection;
+
+use crate::meta::render_value;
+use crate::parameter::InputOutput;
+use crate::parameter::Parameter;
+
+/// A map of metadata key-value pairs, sorted by key.
+type MetaMap = BTreeMap<String, MetadataValue>;
+
+/// A group of inputs.
+#[derive(Debug, Eq, PartialEq)]
+pub struct Group(pub String);
+
+impl PartialOrd for Group {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Group {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        if self.0 == "Common" {
+            return std::cmp::Ordering::Less;
+        }
+        if other.0 == "Common" {
+            return std::cmp::Ordering::Greater;
+        }
+        if self.0 == "Resources" {
+            return std::cmp::Ordering::Greater;
+        }
+        if other.0 == "Resources" {
+            return std::cmp::Ordering::Less;
+        }
+        self.0.cmp(&other.0)
+    }
+}
+
+/// A callable (workflow or task) in a WDL document.
+pub trait Callable {
+    /// Get the name of the callable.
+    fn name(&self) -> &str;
+
+    /// Get the [`MetaMap`] of the callable.
+    fn meta(&self) -> &MetaMap;
+
+    /// Get the inputs of the callable.
+    fn inputs(&self) -> &[Parameter];
+
+    /// Get the outputs of the callable.
+    fn outputs(&self) -> &[Parameter];
+
+    /// Get the description of the callable.
+    fn description(&self) -> Markup {
+        self.meta()
+            .get("description")
+            .map(render_value)
+            .unwrap_or_else(|| html! {})
+    }
+
+    /// Get the required input parameters of the callable.
+    fn required_inputs(&self) -> impl Iterator<Item = &Parameter> {
+        self.inputs().iter().filter(|param| {
+            param
+                .required()
+                .expect("inputs should return Some(required)")
+        })
+    }
+
+    /// Get the sorted set of unique `group` values of the inputs.
+    ///
+    /// The `Common` group, if present, will always be first in the set,
+    /// followed by any other groups in alphabetical order, and lastly
+    /// the `Resources` group.
+    fn input_groups(&self) -> BTreeSet<Group> {
+        self.inputs()
+            .iter()
+            .filter_map(|param| param.group())
+            .map(|arg0: Group| Group(arg0.0.clone()))
+            .collect()
+    }
+
+    /// Get the inputs of the callable that are part of `group`.
+    fn inputs_in_group<'a>(&'a self, group: &'a Group) -> impl Iterator<Item = &'a Parameter> {
```

Comment:
For this and the next method, there's actually a cool method I wasn't aware of called [`partition()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition) that can partition the inputs into "grouped" or "non-grouped" in one call. I'd consider using that.

Reply:
I'm unclear if this is easy with what's going on with required. So it's more just a thing to consider.